File: mud-platform-client/src/lib/components/Input/CommandParser.ts
export type ActionType =
    | 'north' | 'northeast' | 'east' | 'southeast' | 'south' | 'southwest' | 'west' | 'northwest' | 'up' | 'down'
    | 'open' | 'enter' | 'look' | 'say' | 'whisper' | 'tell' | 'who'
    | 'take' | 'drop' | 'attack' | 'talk' | 'inventory' | 'craft' | 'use' | 'unknown';
export type Direction = 'N' | 'S' | 'E' | 'W' | 'NE' | 'NW' | 'SE' | 'SW' | 'UP' | 'DOWN';

export interface ParsedCommand {
    action: ActionType;
    target?: string;
    // direction?: Direction; // Deprecated
    quantity?: number;
    items?: string[];
    message?: string;
    recipient?: string;
    raw: string;
    confidence: number; // 0.0 to 1.0
}

export interface ParserContext {
    lastTarget?: string;
    lastNPC?: string;
    lastItem?: string;
}

export class CommandParser {
    private commandAliases: Map<string, string[]>;
    private contextMemory: ParserContext;

    constructor(context: ParserContext = {}) {
        this.contextMemory = context;
        this.commandAliases = new Map([
            ['north', ['n']],
            ['northeast', ['ne']],
            ['east', ['e']],
            ['southeast', ['se']],
            ['south', ['s']],
            ['southwest', ['sw']],
            ['west', ['w']],
            ['northwest', ['nw']],
            ['up', ['u']],
            ['down', ['d', 'dn']],
            ['open', []],
            ['enter', ['go in', 'step through']],
            ['look', ['l', 'examine', 'inspect', 'view']],
            ['say', ['speak']],
            ['whisper', ['psst']],
            ['tell', ['message', 'msg', 'pm']],
            ['who', ['players', 'online']],
            ['take', ['get', 'grab', 'pick', 'pickup']],
            ['drop', ['release', 'discard', 'throw']],
            ['attack', ['hit', 'fight', 'strike', 'kill']],
            ['talk', ['chat']],
            ['inventory', ['inv', 'i', 'items', 'bag']],
            ['craft', ['make', 'create', 'build', 'forge']],
            ['use', ['consume', 'activate', 'apply']],
        ]);
    }

    updateContext(context: Partial<ParserContext>) {
        this.contextMemory = { ...this.contextMemory, ...context };
    }

    parse(input: string): ParsedCommand {
        // Normalize input
        const normalized = input.trim(); // Don't lowercase yet to preserve message case

        // Check for empty input
        if (!normalized) {
            return this.createError('Empty command');
        }

        // Check for quoted speech
        const quoted = this.parseQuotedSpeech(normalized);
        if (quoted) return quoted;

        // Lowercase for command matching
        const lowerInput = normalized.toLowerCase();

        // Try exact command match first
        const exactMatch = this.tryExactMatch(lowerInput, normalized);
        if (exactMatch) return exactMatch;

        // Try natural language parsing
        const nlpMatch = this.parseNaturalLanguage(lowerInput, normalized);
        if (nlpMatch.confidence > 0.6) return nlpMatch;

        // Try fuzzy matching for typos
        const fuzzyMatch = this.tryFuzzyMatch(lowerInput);
        if (fuzzyMatch.confidence > 0.5) return fuzzyMatch;

        // Unknown command
        return this.createError(`Unknown command: "${input}". Type "help" for commands.`);
    }

    private createError(message: string): ParsedCommand {
        return {
            action: 'unknown',
            raw: message,
            confidence: 0.0
        };
    }

    private parseQuotedSpeech(input: string): ParsedCommand | null {
        if (input.startsWith('"') || input.startsWith("'")) {
            const message = input.substring(1).trim();
            return {
                action: 'say',
                message,
                raw: input,
                confidence: 1.0
            };
        }
        return null;
    }

    private tryExactMatch(lowerInput: string, originalInput: string): ParsedCommand | null {
        const words = lowerInput.split(' ');
        const command = words[0];
        const args = words.slice(1).join(' ');

        // Get original args to preserve case for messages
        const originalWords = originalInput.split(' ');
        const originalArgs = originalWords.slice(1).join(' ');

        for (const [action, aliases] of this.commandAliases) {
            if (action === command || aliases.includes(command)) {
                return this.parseActionArgs(action as ActionType, args, originalArgs, originalInput);
            }
        }

        return null;
    }

    private parseActionArgs(action: ActionType, args: string, originalArgs: string, raw: string): ParsedCommand {
        const result: ParsedCommand = { action, raw, confidence: 1.0 };

        switch (action) {
            case 'say':
                result.message = originalArgs;
                break;
            case 'whisper':
                // Format: whisper <recipient> <message>
                const whisperParts = originalArgs.split(' ');
                if (whisperParts.length >= 2) {
                    result.recipient = whisperParts[0];
                    result.message = whisperParts.slice(1).join(' ');
                }
                break;
            case 'tell':
                // Format: tell <recipient> <message>
                const tellParts = originalArgs.split(' ');
                if (tellParts.length >= 2) {
                    result.recipient = tellParts[0];
                    result.message = tellParts.slice(1).join(' ');
                }
                break;
            case 'open':
            case 'enter':
            case 'take':
            case 'craft':
            case 'drop':
            case 'use':
            case 'look':
                result.target = this.cleanArgs(args);
                break;
            case 'attack':
                result.target = this.cleanArgs(args) || this.contextMemory.lastTarget;
                break;
            case 'talk':
                result.target = this.cleanArgs(args) || this.contextMemory.lastNPC;
                break;
        }

        return result;
    }

    private cleanArgs(args: string): string | undefined {
        if (!args) return undefined;

        // Remove common prepositions/articles that might be left in args
        // e.g. "at sword", "to merchant", "up the sword"
        let cleaned = args
            .replace(/^(at|to|with|up|down|the|a|an)\s+/i, '')
            .replace(/^(the|a|an)\s+/i, ''); // Run again for "up the" -> "the" -> ""

        return cleaned.trim() || undefined;
    }

    private parseNaturalLanguage(input: string, originalInput: string): ParsedCommand {
        // Movement patterns: "go north", "walk east"
        if (/go|walk|move|head|travel/.test(input)) {
            for (const [action, aliases] of this.commandAliases) {
                if (['north', 'northeast', 'east', 'southeast', 'south', 'southwest', 'west', 'northwest', 'up', 'down'].includes(action)) {
                    // Check if direction is in input
                    if (input.includes(action) || aliases.some(a => input.includes(` ${a} `) || input.endsWith(` ${a}`))) {
                        return {
                            action: action as ActionType,
                            raw: originalInput,
                            confidence: 0.9,
                        };
                    }
                }
            }
        }

        // Taking items: "pick up the sword", "get iron ore"
        if (/pick up|take|get|grab/.test(input)) {
            const item = this.extractItemName(input);
            return {
                action: 'take',
                target: item,
                raw: originalInput,
                confidence: 0.85,
            };
        }

        // Dropping items
        if (/drop|throw away|discard/.test(input)) {
            const item = input.replace(/^(drop|throw away|discard)\s+/i, '').replace(/^(the|a|an)\s+/i, '').trim();
            return {
                action: 'drop',
                target: item,
                raw: originalInput,
                confidence: 0.85,
            };
        }

        // Talking
        if (/talk|speak|chat/.test(input)) {
            const target = this.extractTarget(input);
            return {
                action: 'talk',
                target: target || this.contextMemory.lastNPC,
                raw: originalInput,
                confidence: target ? 0.85 : 0.6,
            };
        }

        // Looking: "look at fountain", "examine the door"
        if (/look|examine|inspect/.test(input)) {
            const target = this.extractTarget(input);
            return {
                action: 'look',
                target,
                raw: originalInput,
                confidence: 0.8,
            };
        }

        return { action: 'unknown', raw: originalInput, confidence: 0.0 };
    }

    private extractItemName(input: string): string {
        // Remove command words
        let cleaned = input
            .replace(/^(pick up|take|get|grab|make|craft|create|forge)\s+/i, '')
            .replace(/^(the|a|an)\s+/i, '');

        return cleaned.trim();
    }

    private extractTarget(input: string): string | undefined {
        // Remove command words and articles
        let cleaned = input
            .replace(/^(talk|speak|chat|attack|fight|hit|look|examine)\s+/i, '')
            .replace(/^(to|with|at|the|a|an)\s+/i, '');

        return cleaned.trim() || undefined;
    }

    private tryFuzzyMatch(input: string): ParsedCommand {
        const words = input.split(' ');
        const firstWord = words[0];

        let bestMatch: string | null = null;
        let bestDistance = Infinity;

        for (const [command, aliases] of this.commandAliases) {
            const allVariants = [command, ...aliases];

            for (const variant of allVariants) {
                const distance = this.levenshteinDistance(firstWord, variant);

                if (distance < bestDistance && distance <= 2) {
                    bestDistance = distance;
                    bestMatch = command;
                }
            }
        }

        if (bestMatch) {
            // Reconstruct command with corrected first word
            const correctedInput = [bestMatch, ...words.slice(1)].join(' ');
            return this.parse(correctedInput);
        }

        return { action: 'unknown', raw: input, confidence: 0.0 };
    }

    private levenshteinDistance(str1: string, str2: string): number {
        const matrix: number[][] = [];

        for (let i = 0; i <= str2.length; i++) {
            matrix[i] = [i];
        }

        for (let j = 0; j <= str1.length; j++) {
            matrix[0][j] = j;
        }

        for (let i = 1; i <= str2.length; i++) {
            for (let j = 1; j <= str1.length; j++) {
                if (str2.charAt(i - 1) === str1.charAt(j - 1)) {
                    matrix[i][j] = matrix[i - 1][j - 1];
                } else {
                    matrix[i][j] = Math.min(
                        matrix[i - 1][j - 1] + 1,
                        matrix[i][j - 1] + 1,
                        matrix[i - 1][j] + 1
                    );
                }
            }
        }

        return matrix[str2.length][str1.length];
    }
}

File: tw-backend/internal/game/processor/command_parser.go
package processor

import (
	"regexp"
	"strings"

	"tw-backend/cmd/game-server/websocket"
)

// CommandParser parses raw text commands into structured CommandData
type CommandParser struct {
	// Command aliases map
	aliases map[string][]string
}

// NewCommandParser creates a new command parser
func NewCommandParser() *CommandParser {
	return &CommandParser{
		aliases: map[string][]string{
			"north":     {"n"},
			"northeast": {"ne"},
			"east":      {"e"},
			"southeast": {"se"},
			"south":     {"s"},
			"southwest": {"sw"},
			"west":      {"w"},
			"northwest": {"nw"},
			"up":        {"u"},
			"down":      {"d", "dn"},
			"look":      {"l", "examine", "inspect", "view", "ex"},
			"say":       {"speak"},
			"whisper":   {"psst"},
			"tell":      {"message", "msg", "pm"},
			"who":       {"players", "online"},
			"get":       {"take", "grab", "pick", "pickup"},
			"push":      {"pull", "move"},
			"drop":      {"release", "discard", "throw"},
			"attack":    {"hit", "fight", "strike", "kill"},
			"talk":      {"chat"},
			"inventory": {"inv", "i", "items", "bag"},
			"craft":     {"make", "build", "forge"},
			"use":       {"consume", "activate", "apply"},
			"reply":     {"r"},
			"lobby":     {"exit", "leave", "hub"},
			"create":    nil,
			"weather":   {"climate", "forecast"},
			"ecosystem": {"eco"},
			"world":     nil,
			"fly":       nil,
		},
	}
}

// ParseText parses raw text input into a CommandData struct
func (p *CommandParser) ParseText(text string) *websocket.CommandData {
	text = strings.TrimSpace(text)
	if text == "" {
		return nil
	}

	// Split into words
	words := strings.Fields(text)
	if len(words) == 0 {
		return nil
	}

	// Get the command (first word)
	commandWord := strings.ToLower(words[0])
	args := words[1:]

	// Resolve command alias
	action := p.resolveAlias(commandWord)

	// Build CommandData based on action type
	cmd := &websocket.CommandData{Action: action}

	switch action {
	case "say":
		// Everything after "say" is the message
		if len(args) > 0 {
			message := strings.Join(args, " ")
			cmd.Message = &message
		}

	case "whisper", "tell":
		// Format: whisper/tell <recipient> <message>
		if len(args) >= 2 {
			recipient := args[0]
			message := strings.Join(args[1:], " ")
			cmd.Recipient = &recipient
			cmd.Message = &message
		} else if len(args) == 1 {
			// Just recipient, no message (error will be handled by processor)
			recipient := args[0]
			cmd.Recipient = &recipient
		}

	case "reply":
		// Format: reply <message>
		// Recipient will be resolved by the processor using last tell sender
		if len(args) > 0 {
			message := strings.Join(args, " ")
			cmd.Message = &message
		}

	case "create":
		// Format: create world, create <target>
		if len(args) > 0 {
			target := strings.Join(args, " ")
			cmd.Target = &target
		}

	case "enter":
		// Format: enter <world_id/portal_name/target>
		if len(args) > 0 {
			target := strings.Join(args, " ")
			cmd.Target = &target
		}

	case "look", "get", "push", "drop", "attack", "talk", "craft", "use", "open", "face":
		// Format: <action> <target>
		// Join all args as target (handles multi-word targets like "iron sword")
		if len(args) > 0 {
			target := strings.Join(args, " ")
			// Remove common articles/prepositions
			target = p.cleanTarget(target)
			cmd.Target = &target
		}

	// Direction commands don't need additional processing
	case "north", "northeast", "east", "southeast", "south", "southwest", "west", "northwest", "up", "down":
		if len(args) > 0 {
			target := strings.Join(args, " ")
			cmd.Target = &target
		}

	// Commands without arguments
	case "who", "inventory", "lobby":
		// No additional fields needed

	case "help":
		// Format: help [args]
		if len(args) > 0 {
			target := strings.Join(args, " ")
			cmd.Target = &target
		}

	case "watcher":
		// Format: watcher <world_id>
		if len(args) > 0 {
			target := strings.Join(args, " ")
			cmd.Target = &target
		}

	case "ecosystem":
		// Format: ecosystem <subcommand> <args>
		// e.g. ecosystem spawn rabbit -> Target="spawn", Message="rabbit"
		if len(args) >= 2 {
			target := args[0]
			message := strings.Join(args[1:], " ")
			cmd.Target = &target
			cmd.Message = &message
		} else if len(args) == 1 {
			target := args[0]
			cmd.Target = &target
		}

	case "world":
		// Format: world <subcommand> <args>
		// e.g. world simulate 1000000 -> Target="simulate", Message="1000000"
		if len(args) >= 2 {
			target := args[0]
			message := strings.Join(args[1:], " ")
			cmd.Target = &target
			cmd.Message = &message
		} else if len(args) == 1 {
			target := args[0]
			cmd.Target = &target
		}

	case "fly":
		// Format: fly <height>
		if len(args) >= 1 {
			target := strings.Join(args, " ")
			cmd.Target = &target
		}

	default:
		// Unknown command - keep the action as-is
		// Error will be handled by processor
	}

	return cmd
}

// resolveAlias converts an alias to its canonical action name
func (p *CommandParser) resolveAlias(word string) string {
	// Check if it's already a known action
	for action, aliases := range p.aliases {
		if word == action {
			return action
		}
		// Check aliases
		for _, alias := range aliases {
			if word == alias {
				return action
			}
		}
	}

	// Not found - return as-is (will be handled as unknown command)
	return word
}

// cleanTarget removes common articles and prepositions from targets
func (p *CommandParser) cleanTarget(target string) string {
	target = strings.ToLower(target)

	// Remove leading articles/prepositions
	re := regexp.MustCompile(`^(at|to|with|up|down|the|a|an)\s+`)
	target = re.ReplaceAllString(target, "")

	// Run again for cases like "up the sword" -> "the sword" -> "sword"
	re = regexp.MustCompile(`^(the|a|an)\s+`)
	target = re.ReplaceAllString(target, "")

	return strings.TrimSpace(target)
}

// SimulationConfig holds parsed simulation command arguments
type SimulationConfig struct {
	Years            int64
	Epoch            string
	Goal             string
	WaterLevel       string
	SimulateGeology  bool
	SimulateLife     bool
	SimulateDiseases bool
}

// ParseSimulationArgs parses simulation command arguments into a config struct.
// RED STATE: Returns nil - not yet implemented. Tests should fail.
func ParseSimulationArgs(argsStr string) *SimulationConfig {
	// TODO: Implement proper parsing
	// This is intentionally returning nil to fail tests (TDD Red state)
	return nil
}
